vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoClientImpl.java

```diff
@@ -668,6 +690,64 @@ public MongoClient dropIndex(String collection, String indexName, Handler<AsyncR
     return this;
   }
 
+  @Override
+  public MongoClient watch(String collection, JsonArray pipeline, WatchOptions options, Handler<AsyncResult<MongoClientChangeStream<MongoClientChange>>> resultHandler) {
+    return createWatchIterable(collection, pipeline, options, false, resultHandler);
+  }
+
+  @Override
+  public MongoClient watchReplaceRoot(String collection, JsonArray pipeline, WatchOptions options, Handler<AsyncResult<MongoClientChangeStream<JsonObject>>> resultHandler) {
+    throw new UnsupportedOperationException();
+    // Not supported until https://jira.mongodb.org/browse/JAVA-2828 is fixed
+    // return createWatchIterable(collection, pipeline, options, true, resultHandler);
+  }
+  
+  @SuppressWarnings("unchecked")
+  private <T> MongoClient createWatchIterable(String collection, JsonArray pipeline, WatchOptions options, boolean replacedRootDocument, Handler<AsyncResult<MongoClientChangeStream<T>>> resultHandler) {
+    requireNonNull(collection);
+    requireNonNull(pipeline);
+
+    if (options == null) {
+      options = new WatchOptions();
+    } else {
+      if (options.getFullDocument() != null
+        && !options.getFullDocument().equals(WatchOptions.FULL_DOCUMENT_UPDATE_LOOKUP)) {
+        resultHandler.handle(Future.failedFuture(new IllegalArgumentException("fullDocument must be null or \"updatedLookup\"")));
+        return this;
+      }
+    }
+
+    try {
+      List<Bson> pipelineBson = IntStream.range(0, pipeline.size())
+        .mapToObj(pipeline::getJsonObject)
+        .map(JsonObjectBsonAdapter::new)
+        .collect(Collectors.toList());
+
+      ChangeStreamIterable<JsonObject> view = getCollection(collection)
+        .watch(pipelineBson)
+        .fullDocument((options.getFullDocument() != null && options.getFullDocument().equals(WatchOptions.FULL_DOCUMENT_UPDATE_LOOKUP)) ? FullDocument.UPDATE_LOOKUP : FullDocument.DEFAULT)
+        .batchSize(options.getBatchSize());
+      if (options.getResumeAfter() != null) {
+        view.resumeAfter((BsonDocument) toBson(options.getResumeAfter()));
+      }
+      if (options.getMaxAwaitTimeMS() != WatchOptions.DEFAULT_MAX_AWAIT_TIME_MS) {
+        view.maxAwaitTime(options.getMaxAwaitTimeMS(), TimeUnit.MILLISECONDS);
+      }
+
+      Context context = vertx.getOrCreateContext();
+      if (replacedRootDocument) {
+        MongoIterableStream<JsonObject> adaptor = new MongoIterableStream<>(view.withDocumentClass(JsonObject.class), context, options.getBatchSize());
+        resultHandler.handle(Future.succeededFuture((MongoClientChangeStream<T>)new MongoClientReplacedRootStreamImpl(adaptor)));
```

Comment:
I think this casting may show a mistake in how control is handled here, `createWatchIterable` could return `ChangeStreamIterable<JsonObject>` then let the upstream methods deal with converting it to either a `MongoClientChangeStream<JsonObject>` or `MongoClientChangeStream<MongoClientChange>`